Project/DAQ_System/Program/Src/Platform/I2C/STM/i2cHalWrapper_stmf4.cpp

```diff
@@ -0,0 +1,30 @@
+/*
+ * i2cHalWrapper_stmf4.cpp
+ *
+ *  Created on: Mar 4, 2024
+ *      Author: Nico
+ */
+#include "i2cHalWrapper_stmf4.hpp"
+
+namespace platform {
+
+	i2cHalWrapperStmf4::i2cHalWrapperStmf4(I2C_HandleTypeDef &hi2c, uint8_t SAD)
+			: bx_i2c_(hi2c) {
+		slave_add_ = SAD;
+	}
+
+	i2cHalWrapperStmf4::~i2cHalWrapperStmf4(){}
+
+	void i2cHalWrapperStmf4::Master_Transmit(uint8_t *commands){
+		HAL_I2C_Master_Transmit(&bx_i2c_,slave_add_, commands, kMaxBytes, 0xFFFFFFFF);
+	}
+	void i2cHalWrapperStmf4::Master_Receive(uint8_t *response){
+		HAL_I2C_Master_Receive(&bx_i2c_, slave_add_, response, response_size_, 0xFFFFFFFF);
```

Comment:
Receiving for I2C is kinda weird since it actually requires both a transmit and a receive function.  if you see how I implemented a receive you will notice that I first transmit then receive. It seems counter-intuitive to transmit in a receive function, but the whole purpose is to first transmit the register address, and then you can receive data from that register.

![image](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/eb0ad2e7-1ee8-4181-869c-70ad0053404a)

Note: the if statements are just checking that the functions HAL_I2C_Master_Transmit and HAL_I2C_Master_Receive are properly functioning. If it doesn't then receive function outputs are FALSE and cancel the whole function.

Reply:
That *might* be device-specific. In the context of SPI communication, I have seen a separate function along the lines of `WriteRead(tx_buffer[], rx_buffer[])` that do exactly as you mentioned.

Reply:
I am relatively sure that the way the Will implemented the receiving is universal to i2 communications